### PR TITLE
[US16] Create endpoints for registration/login.

### PR DESCRIPTION
**Pull request by BukanNadya**
Сhange bean BCryptPasswordEncoder on MyDelegatedEncoder. In the confirmation email we send an email with a code, which we then compare with the cache.

### CODE REVIEW

File: social_network_api/src/main/java/com/danit/socialnetwork/MyDelegatedEncoder.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.crypto.password.DelegatingPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+import java.util.HashMap;
+
+@Configuration
+public class MyDelegatedEncoder {
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/MyDelegatedEncoder.java:11]: Do we really need this?

File: social_network_api/src/main/java/com/danit/socialnetwork/config/GuavaCache.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  public Integer getUnchecked(String key) {
+    return cache.getIfPresent(key);
+  }
+
+}
+
+
+
+
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/config/GuavaCache.java:42]: Please leave only 1 line at the end of file.

File: social_network_api/src/main/java/com/danit/socialnetwork/config/JavaMailSenderConfig.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    mailSender.setHost(host);
+    mailSender.setPort(port);
+
+    mailSender.setUsername(username);
+    mailSender.setPassword(password);
+
+    Properties props = mailSender.getJavaMailProperties();
+    props.put("mail.smtp.auth", auth);
+    props.put("mail.smtp.starttls.enable", enable);
+    props.put("mail.debug", debug);
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/config/JavaMailSenderConfig.java:42]: Do we need this?

File: social_network_api/src/main/java/com/danit/socialnetwork/dto/ActivateCodeRequest.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.danit.socialnetwork.dto;
+
+import lombok.Data;
+
+@Data
+public class ActivateCodeRequest {
+  Integer code;
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/dto/ActivateCodeRequest.java:7]: Please, add private access modifier.

File: social_network_api/src/main/java/com/danit/socialnetwork/config/GuavaCache.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      })
+      .maximumSize(20)
+      .expireAfterWrite(5, TimeUnit.MINUTES)
+      .build();
+
+  public void put(String key, Integer value) {
+    cache.put(key, value);
+  }
+
+  public Integer getUnchecked(String key) {
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/config/GuavaCache.java:34]: Please, make these 2 methods static.

File: social_network_api/src/main/java/com/danit/socialnetwork/rest/UserRestController.java [hunk 6]
```diff
@@ -1,28 +1,110 @@
...
+    dbUser.setEmail(request.getEmail());
+    dbUser.setName(request.getName());
+    dbUser.setDateOfBirth(dateOfBirth);
+
+    Map<String, String> response = new HashMap<>();
+    if (userService.save(dbUser)) {
+      response.put("registration", "true");
+    } else {
+      response.put("registration", "false");
+    }
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/rest/UserRestController.java:56]: You can just make response.put("registration", userService.save(dbUser))

File: social_network_api/src/main/java/com/danit/socialnetwork/rest/UserRestController.java [hunk 7]
```diff
@@ -1,28 +1,110 @@
...
+      @RequestBody ActivateCodeRequest request) {
+    Integer code = request.getCode();
+    boolean isActivated = userService.activateUser(code);
+    Map<String, String> response = new HashMap<>();
+
+    if (isActivated) {
+      response.put("activate", "true");
+    } else {
+      response.put("activate", "false");
+    }
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/rest/UserRestController.java:102]: You can just make response.put("activate", isActivated);

File: social_network_api/src/main/java/com/danit/socialnetwork/security/WebSecurityConfig.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  @Bean
+  @Override
+  public AuthenticationManager authenticationManagerBean() throws Exception {
+    return super.authenticationManagerBean();
+  }
+
+  @Bean
+  GuavaCache guavaCache() {
+    return new GuavaCache();
+  }
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/security/WebSecurityConfig.java:34]: You can remove this bean and use static methods.

File: social_network_api/src/main/java/com/danit/socialnetwork/service/MailSenderImpl.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    mailSender.send(mailMessage);
+  }
+
+  @Override
+  public void send(SimpleMailMessage simpleMessage) throws MailException {
+
+  }
+
+  @Override
+  public void send(SimpleMailMessage... simpleMessages) throws MailException {
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/service/MailSenderImpl.java:36]: Can you remove this methods? Or they can throw UnsuportedOperationException

File: social_network_api/src/main/java/com/danit/socialnetwork/service/UserService.java [hunk 10]
```diff
@@ -3,11 +3,22 @@
...
+
+  Optional<DbUser> findByUsername(String username);
+
+  public boolean activateUser(Integer code);
+
+
+  boolean save(DbUser dbUser);
+
+
+  boolean sendLetter(String name, String email);
```
Comment (miegami) [inline social_network_api/src/main/java/com/danit/socialnetwork/service/UserService.java:22]: A lot of empty lines

File: social_network_api/src/main/resources/application.yaml [hunk 11]
```diff
@@ -3,17 +3,20 @@ server:
 
 spring:
   datasource:
-    url: jdbc:h2:file:D:\Java\projects\FP4SocialNetwork\social_network_api\src\main\resources\data\tweeterDB;AUTO_SERVER=true
+#    url: jdbc:h2:mem:tweeterDB
```
Comment (miegami) [inline social_network_api/src/main/resources/application.yaml:6]: Please, remove commented line.

**Review by AlexKhmarenko — APPROVED** [review]

File: social_network_api/src/main/java/com/danit/socialnetwork/MyDelegatedEncoder.java (hunk 1 above)
Comment (BukanNadya) [reply social_network_api/src/main/java/com/danit/socialnetwork/MyDelegatedEncoder.java:11]: I change this on BCryptPasswordEncoder()

**Review by miegami — APPROVED** [review]

**Review by miegami — APPROVED** [review]

**Pull request merged**